drivers: smbus: stm32: add support for pcall api

cfriedt opened the pull request:
Add SMBus Process Call / Process Call Block APIs for STM32.

This implementation also supports PEC mode (packet error checking) and is dependent on PEC support already being supported in-tree.

For more details, please see Sections 6.5.6 and 6.5.8 of the SMBus 3.1 Specification.
https://smbus.org/specs/SMBus_3_1_20180319.pdf

Commit:
drivers: smbus: stm32: implement smbus_stm32_block_pcall() with pec

Add SMBus Block Write-Block Read Process Call API for STM32.

This implementation also supports PEC mode (packet error checking) and is
dependent on PEC support already being supported in-tree.

Signed-off-by: James Growden <jgrowden@tenstorrent.com>
Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

etienne-lms reviewed (commented):
Commit "drivers: smbus: stm32: add support for pcall api" LGTM.

benediktibk commented on drivers/smbus/smbus_stm32.c:
```diff
@@ -288,6 +288,70 @@ static int smbus_stm32_block_read(const struct device *dev, uint16_t periph_addr
+	 * This works because the STM I2C driver processes each message serially.
+	 * The addressing math assumes little-endian.
+	 */
+	msgs[3].buf = (uint8_t *)&msgs[4].len;
```
neat trick

cfriedt replied:
I think that came from @jgrowdenTT 😀

jgrowdenTT replied:
Thanks, but I can't claim credit :) it's a repeat of the idea in `smbus_stm32_block_read`

cfriedt replied:
Probably @alewycky-tenstorrent in that case 😀

benediktibk reviewed (changes requested):
In the SMBus specification the process call seems to be specified differently:
<img width="832" height="674" alt="grafik" src="https://github.com/user-attachments/assets/4e20a3e3-c5dc-4531-b656-9ba215da7d8a" />

There is no count for send and receive with a variable length, just a hard coded word exchange (write and read)?

There seems to be no difference between version 3.1 and 3.2 of the SMBus specification regarding the process call, btw.

etienne-lms commented:
> In the SMBus specification the process call seems to be specified differently:
(...)
There is no count for send and receive with a variable length, just a hard coded word exchange (write and read)?

I see this change implements the block write/read process call (Section "6.5.8 Block Write-Block Read Process Call" of the spec).

Indeed the "Process Call" (single word) this is already implemented by `smbus_stm32_pcall()` handler but currently does not support PEC. Commit message (and P-R description) should be fixed.

@cfriedt, @jgrowdenTT, did you plan to also implement single word "Process Call"?

Commit:
drivers: smbus: stm32: implement smbus_stm32_pcall() with pec

Implement smbus_stm32_pcall() with packet error correction.

For more information, please see chapter 6.5.6 of the SMBus specification.
https://smbus.org/specs/SMBus_3_1_20180319.pdf

Signed-off-by: Chris Friedt <cfriedt@tenstorrent.com>

cfriedt commented:
@etienne-lms @benediktibk @mathieuchopstm @erwango - please take another look

mathieuchopstm reviewed (commented):
Otherwise LGTM

cfriedt commented on drivers/smbus/smbus_stm32.c:
```diff
@@ -346,18 +345,46 @@ static int smbus_stm32_word_data_read(const struct device *dev, uint16_t periph_
+			.flags = I2C_MSG_WRITE,
+		},
+		{
+			.buf = recv_word,
```
All stm32 are little endian. If this driver were ever be made to be completely generic, I could see doing that in the future though.

benediktibk replied:
Okay, looks like the ARM endianness configuration feature is always set to little on the STM32.

benediktibk commented on drivers/smbus/smbus_stm32.c:
```diff
@@ -346,18 +345,46 @@ static int smbus_stm32_word_data_read(const struct device *dev, uint16_t periph_
+			.flags = I2C_MSG_WRITE,
+		},
+		{
+			.buf = &send_word,
```
Okay, looks like the ARM endianness configuration feature is always set to little on the STM32.
